### DISCUSSION TITLE
Cannot make `--yarnrc` option to work in GitLab CI

### DISCUSSION THREAD

**Discussion by p3k** [renovatebot/renovate#9331] [Request Help]
**TL;DR:** Renovate does not find the Yarn binary we try to enforce using `yarn-path` with the `--yarnrc` flag when running it via GitLab CI.

We are running Renovate as a scheduled GitLab CI job:

```yml
dependencies:
  stage: update
  cache: {}
  variables:
    LOG_LEVEL: debug
    RENOVATE_CONFIG_FILE: ./.gitlab/renovate.json
  before_script:
    - ./gradlew :yarn_global_add_renovate
    - ln -s /builds/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn /usr/bin/yarn
    - echo $HOME
    - echo $PWD
    - echo $yarn_path
  script:
    - echo "Using Yarn version $(yarn --version) in $(which yarn)"
    - echo "Using Renovate version $(renovate --version) in $(which renovate)"
    - >
      renovate
      --yarnrc "yarn-path '/builds/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn'"
      --token ${GITLAB_AUTH_TOKEN}
  only:
    - schedules
```

Breaking it down:

1. In `before_script` we install Renovate globally using Gradle’s Node / Yarn plugin
2. This creates the path `/builds/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn` because we enforce this version in `build.gradle`:
   ```gradle
   node {
     yarnVersion = '1.22.10'
   }
   ```
   Furthermore, in our top-level `.yarnrc` we require the Yarn binary of this very version – this is *not* the correct path for the CI job, but we need it exactly like this for local development purposes:
   ```plain
   yarn-path "./.gradle/yarn/yarn-v1.22.10/bin/yarn"
   ```
3. We then symlink the 1.22.10 version of the binary to `/usr/bin/yarn`
3. The `echo` commands (for debugging) show:
   ```sh
   /root # $HOME
   /builds/portal/portal # $PWD
   /builds/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn # $yarn_path
   ```
4. Now `script` starts, first outputting some more debug messages, looking good:
   ```plain
   Using Yarn version 1.22.10 in /usr/bin/yarn
   Using Renovate version 24.98.1 in /usr/local/bin/renovate
   ```
5. Finally, Renovate is run with all necessary tokens and (important for this question) with the `--yarnrc` flag – which as far as I understand should overwrite our own settings in `.yarnrc` (which contains a *different* Yarn path, see above) with the desired path to the Yarn binary.
6. Debug mode is enabled so I see a lot of information in the CI job output, most significant being
    ```plain
    DEBUG: lock file error (repository=portal/portal, branch=renovate/eslint-config-prettier-8.x)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts",
         "stdout": "",
         "stderr": "internal/modules/cjs/loader.js:818\n  throw err;\n  ^\n\nError: Cannot find module '/tmp/renovate/repos/gitlab/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn'\n    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)\n    at Function.Module._load (internal/modules/cjs/loader.js:667:27)\n    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)\n    at internal/main/run_main_module.js:17:47 {\n  code: 'MODULE_NOT_FOUND',\n  requireStack: []\n}\n",
         "message": "Command failed: yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\ninternal/modules/cjs/loader.js:818\n  throw err;\n  ^\n\nError: Cannot find module '/tmp/renovate/repos/gitlab/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn'\n    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)\n    at Function.Module._load (internal/modules/cjs/loader.js:667:27)\n    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)\n    at internal/main/run_main_module.js:17:47 {\n  code: 'MODULE_NOT_FOUND',\n  requireStack: []\n}\n",
         "stack": "Error: Command failed: yarn install --ignore-engines --ignore-platform --network-timeout 100000 --ignore-scripts\ninternal/modules/cjs/loader.js:818\n  throw err;\n  ^\n\nError: Cannot find module '/tmp/renovate/repos/gitlab/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn'\n    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)\n    at Function.Module._load (internal/modules/cjs/loader.js:667:27)\n    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12)\n    at internal/main/run_main_module.js:17:47 {\n  code: 'MODULE_NOT_FOUND',\n  requireStack: []\n}\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:314:20)\n    at ChildProcess.EventEmitter.emit (domain.js:483:12)\n    at maybeClose (internal/child_process.js:1022:16)\n    at Socket.<anonymous> (internal/child_process.js:444:11)\n    at Socket.emit (events.js:314:20)\n    at Socket.EventEmitter.emit (domain.js:483:12)\n    at Pipe.<anonymous> (net.js:675:12)"
       },
       "type": "yarn"
    ```
    So Renovate fails to run Yarn for the upgraded package version because it looks for the Yarn binary in `/tmp/renovate/repos/gitlab/portal/portal/.gradle/yarn/yarn-v1.22.10/bin/yarn`.

    Why is Renovate looking there? `/tmp/renovate/repos/gitlab/portal/portal` is Renovate’s temporary path for the checked out branch of the repo, the remaining path seems to come from our original `.yarnrc` file (the local Yarn path, remember?) – which obviously was not overwritten, then… 🤔

    Of course there is no Yarn binary under this path because we did not (and afaik can not) install our desired Yarn binary in the temporary branch checkout.
7. If I remove our `.yarnrc` file suddenly everything is working as expected. Unfortunately, we need the relative path for our local development, and thus, the `.yarnrc` file to make sure every developer is using the same Yarn version.

I simulated the setup locally, doing a `--dry-run` and definitely can see the  `.yarnrc` file in the temporary branch checkout changing (using `watch`) – the only difference being that I call the Renovate binary via a Gradle task (not directly from within the CI script):

```gradle
task checkRenovate {
	doLast {
		exec {
			environment = [
				LOG_LEVEL: 'info',
				RENOVATE_CONFIG_FILE: '.gitlab/renovate.json'
			]

			def gitlabToken = findProperty('net.semanticlab.gitlab.token')

			workingDir rootProject.projectDir
			executable 'renovate'

			args = [
				'--dry-run',
				'--yarnrc', "yarn-path '${rootProject.projectDir}/.gradle/yarn/yarn-v1.22.10/bin/yarn'",
				'--token', gitlabToken
			]
		}
	}
}
```

I feel stuck now and running out of ideas how to even check whether the `.yarnrc` override really works or how to debug any further.

Would be glad for any suggestions – especially if I can avoid migrating to Renovate Docker…

**Comment by rarkins** [comment]
We plan to deprecate the `yarnrc` field in the next major release, so it would be better to step back and understand what's causing you to need the `--yarnrc` flag in the first place. Maybe we have a bug with `.yarnrc` handling, or maybe we've underestimated the need for `yarnrc` overrides.

So the first question is: do you have a `.yarnrc` in the repo already, and if so then can you confirm why you are needing to override it?

Additionally, have you seen our recommended https://gitlab.com/renovate-bot/renovate-runner ?

**Reply by p3k** [reply]
> We plan to deprecate the yarnrc field in the next major release

oh ok looks like using `--yarnrc` in fact is a no-go then, already anyway…

> Maybe we have a bug with .yarnrc handling, or maybe we've underestimated the need for yarnrc overrides.

i think you are onto something here :wink:

> So the first question is: do you have a .yarnrc in the repo already, and if so then can you confirm why you are needing to override it?

yes we have a `.yarnrc` in the repo. we need it for local dev purposes because we want to enforce a specific yarn version. in gitlab’s ci this path is not usable, though. that’s why i am trying to override it with the `--yarnrc` flag. (i think i explained the circumstances quite excessively above.)

> Additionally, have you seen our recommended https://gitlab.com/renovate-bot/renovate-runner ?

yes also know about the docker option but of course for us that would mean a lot of rewriting, refactoring and reconfiguring – and we cannot even be sure our intention to enforce a specifci yarn version can be realized, either.

i guess we are going to try to get rid of our `.yarnrc` file from the repo, seems like the quickest and easiest solution then.

**Reply by p3k** [reply]
for the record: we conclude this by removing `.yarnrc` from version control and instead create it from within a gradle task dependency. this solves our local dev requirement.

regarding the ci job, this works then as already stated due to the symlinked yarn binary. thus, we also removed the `--yarnrc` setting.

@rarkins thanks for your quick reply. i feel quite unbiased regarding the deprecation / removal of the `--yarnrc` option but generally think a way to tell renovate which yarn binary to use would be a good idea.

**Reply by rarkins** [reply]
Per my understanding, we support it fine if you commit the yarn binary to the repo which contains the `.yarnrc` and it references a relative file. I had also thought that was the recommended way from Yarn.

It sounds like your `.yarnrc` referenced a location elsewhere on the system and not within the repo?

What we could maybe do is add logic to detect and strip out non-reachable yarn binary locations in `.yarnrc`. That way you wouldn't get *failures* and wouldn't need to remove the `.yarnrc`, however then you won't get the fixed version being used unless you have `engines.yarn` defined also.

**Reply by p3k** [reply]
> It sounds like your .yarnrc referenced a location not within the repo?

good point, it is created by a gradle task inside the repo but in fact never pushed. 

i also perceive there are some more options configuring yarn, so maybe you really are already doing enough on your side…
